readers.go

```diff
@@ -20,6 +20,9 @@ func ReaderRoot(r io.Reader, h hash.Hash, segmentSize int) (root []byte, err err
 		if readErr == io.EOF {
 			break
 		}
+		if n < segmentSize {
+			segment = segment[:n]
+		}
```

Comment:
this error handling code is a bit confusing now, though it appears to be correct. Here's my suggested replacement:
```go
n, readErr := io.ReadFull(r, segment)
if readErr == io.EOF {
    // All data has been read.
    break
} else if readErr == io.UnexpectedEOF {
    // The last segment is not a full segment, so truncate it.
    // Note that the next ReadFull call will return io.EOF
    segment = segment[:n]
} else {
    return nil, readErr
}
```